features/equal/equal.go

```diff
@@ -215,14 +213,29 @@ func (p *equal) compareBytes(lhs, rhs string, nullable bool) {
 	p.P(`}`)
 }
 
-func (p *equal) compareCall(lhs, rhs string, ccTypeName string, msg *protogen.Message) {
+func (p *equal) compareCall(lhs, rhs string, msg *protogen.Message, nullable bool) {
+	if !nullable {
+		// The p != q check is mostly intended to catch the lhs = nil, rhs = nil case in which we would pointlessly
+		// allocate not just one but two empty values. However, it also provides us with an extra scope to establish
+		// our p and q variables.
+		p.P(`if p, q := `, lhs, `, `, rhs, `; p != q {`)
+		defer p.P(`}`)
+
+		p.P(`if p == nil {`)
+		p.P(`p = &`, p.QualifiedGoIdent(msg.GoIdent), `{}`)
+		p.P(`}`)
+		p.P(`if q == nil {`)
+		p.P(`q = &`, p.QualifiedGoIdent(msg.GoIdent), `{}`)
```

Comment:
Not super happy with allocating values here, but unfortunately it's not very easy to introduce package-private helper variables on the fly for types where this is needed. Also, `nil` values in non-nullable contexts should be an exception.

Reply:
This does not allocate values. They stay in the stack. :)

```
./test_messages_proto2_vtproto.pb.go:1212:9: &TestAllTypesProto2_NestedMessage{} does not escape
./test_messages_proto2_vtproto.pb.go:1215:9: &TestAllTypesProto2_NestedMessage{} does not escape
./test_messages_proto2_vtproto.pb.go:1229:9: &ForeignMessageProto2{} does not escape
./test_messages_proto2_vtproto.pb.go:1232:9: &ForeignMessageProto2{} does not escape
```